src/linux/minidump_writer/mod.rs

```diff
@@ -1,212 +1,268 @@
-    ),
-    #[error("Proc task directory `{0:?}` is not a directory")]
-    ProcPidTaskNotDirectory(String),
-    #[error("Errors while enumerating threads")]
-    EnumerateThreadsErrors(#[source] ErrorList<InitError>),
-    #[error("Failed to enumerate threads")]
-    EnumerateThreadsFailed(#[source] Box<InitError>),
-    #[error("Failed to read process map file")]
-    ReadProcessMapFileFailed(
-        #[source]
-        #[serde(serialize_with = "serialize_proc_error")]
-        ProcError,
-    ),
-    #[error("Failed to aggregate process mappings")]
-    AggregateMappingsFailed(#[source] MapsReaderError),
-    #[error("Failed to enumerate process mappings")]
-    EnumerateMappingsFailed(#[source] Box<InitError>),
+#[derive(Debug)]
+pub struct MinidumpWriter {
+    pub process_id: Pid,
+    threads_suspended: bool,
+    pub threads: Vec<Thread>,
+    pub auxv: AuxvDumpInfo,
+    pub mappings: Vec<MappingInfo>,
+    pub page_size: usize,
+    pub sanitize_stack: bool,
+    pub minidump_size_limit: Option<u64>,
+    pub user_mapping_list: MappingList,
+    pub crashing_thread_context: CrashingThreadContext,
+    stop_timeout: Duration,
+    pub skip_stacks_if_mapping_unreferenced: bool,
+    principal_mapping_address: Option<usize>,
+    pub principal_mapping: Option<MappingInfo>,
+    pub blamed_thread: Pid,
+    pub crash_context: Option<CrashContext>,
+    pub app_memory: AppMemoryList,
+    pub memory_blocks: Vec<MDMemoryDescriptor>,
 }
 
-#[derive(Debug, thiserror::Error, serde::Serialize)]
-pub enum StopProcessError {
-    #[error("Failed to stop the process")]
-    Stop(
-        #[from]
-        #[serde(serialize_with = "serialize_nix_error")]
-        nix::Error,
-    ),
-    #[error("Failed to get the process state")]
-    State(
-        #[from]
-        #[serde(serialize_with = "serialize_proc_error")]
-        ProcError,
-    ),
-    #[error("Timeout waiting for process to stop")]
-    Timeout,
+#[derive(Debug, Clone)]
+pub struct Thread {
+    pub tid: Pid,
+    pub name: Option<String>,
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum ContinueProcessError {
-    #[error("Failed to continue the process")]
-    Continue(#[from] Errno),
+#[derive(Debug, Default)]
+pub enum CrashingThreadContext {
+    #[default]
+    None,
+    CrashContext(MDLocationDescriptor),
+    CrashContextPlusAddress((MDLocationDescriptor, usize)),
 }
 
-/// PTRACE_DETACH the given pid.
-///
-/// This handles special errno cases (ESRCH) which we won't consider errors.
-fn ptrace_detach(child: Pid) -> Result<(), DumperError> {
-    let pid = nix::unistd::Pid::from_raw(child);
-    ptrace::detach(pid, None).or_else(|e| {
-        // errno is set to ESRCH if the pid no longer exists, but we don't want to error in that
-        // case.
-        if e == nix::Error::ESRCH {
-            Ok(())
-        } else {
-            Err(DumperError::PtraceDetachError(child, e))
+impl MinidumpWriterConfig {
+    pub fn new(process_id: Pid, blamed_thread: Pid) -> Self {
```

Comment:
Just a thought for the future: we do this in pretty much all our crates but we should really have a separate `Tid` types for threads. Obviously something for a follow-up.